packages/sanes-chrome-extension/src/logic/user/accountmanager.ts

```diff
@@ -0,0 +1,160 @@
+import {
+  Algorithm,
+  ChainId,
+  publicIdentityEquals,
+  PublicIdentity,
+} from '@iov/bcp';
+import { UserProfile, WalletId } from '@iov/core';
+import { Slip10RawIndex } from '@iov/crypto';
+import { ReadonlyWallet } from '@iov/keycontrol/types/wallet';
+
+export interface AccountInfo {
+  readonly name: string;
+  readonly identities: ReadonlyMap<ChainId, PublicIdentity>;
+}
+
+export interface AccountManagerChainConfig {
+  readonly chainId: ChainId;
+  readonly algorithm: Algorithm;
+  readonly derivePath: (account: number) => ReadonlyArray<Slip10RawIndex>;
+}
+
+export class AccountManager {
+  private readonly _userProfile: UserProfile;
+  private readonly _chains: AccountManagerChainConfig[];
+
+  public constructor(
+    userProfile: UserProfile,
+    chains: ReadonlyArray<AccountManagerChainConfig>
+  ) {
+    this._userProfile = userProfile;
+    this._chains = [...chains];
+  }
+
+  public mnemonic(): string {
+    const firstWalletId = this._userProfile.wallets.value[0].id;
+    // all wallets use the same mnemonic, so we only need to check the first one
+    const mnemonic = this._userProfile.printableSecret(firstWalletId);
+    return mnemonic;
+  }
+
+  public async addChain(newChain: AccountManagerChainConfig): Promise<void> {
+    if (this._chains.find(c => c.chainId === newChain.chainId)) {
+      throw new Error(`Chain with ID ${newChain.chainId} already exists`);
+    }
+
+    this._chains.push(newChain);
+
+    const accountIndices = await this.existingAccountIndices();
+    for (const index of accountIndices) {
+      await this.generateAccount(index);
+    }
+  }
+
+  public async generateAccount(derivation: number): Promise<void> {
+    for (const chain of this._chains) {
+      const { chainId, algorithm, derivePath } = chain;
+      const path = derivePath(derivation);
+      const wallet = this.walletForAlgorithm(algorithm);
+      const identityCreated = await this.identityExistsInProfile(
+        wallet,
+        chainId,
+        path
+      );
+
+      if (!identityCreated) {
+        const identity = await this._userProfile.createIdentity(
+          wallet,
+          chainId,
+          path
+        );
+        await this._userProfile.setIdentityLabel(identity, `${derivation}`);
+      }
+    }
+  }
+
+  public async generateNextAccount(): Promise<void> {
```

Comment:
I would add some documentation here explaining the reasoning for generating naming in accounts.